core/src/main/scala/cromwell/core/retry/IORetry.scala

```diff
@@ -0,0 +1,52 @@
+package cromwell.core.retry
+
+import cats.effect.{IO, Timer}
+
+import scala.concurrent.duration._
+
+object IORetry {
+  def noOpOnRetry[S]: (Throwable, S) => S = (_, s) => s
+
+  /**
+    * When we reach a point where we need to fail the IO (because we ran out of retries, or exception was fatal etc...)
+    * This provides a method to generate a Throwable from the current state and the last failure.
+    */
+  trait StatefulIoError[S] {
+    def toThrowable(state: S, throwable: Throwable): Throwable
+  }
+
+  /**
+    * Same as Retry.withRetry using cats.effect.IO
+    * The state is being passed down upon each retry. It can be use to accumulate information as IO operations
+    * fail and are being retried.
+    */
+  def withRetry[A, S](io: IO[A],
+                      state: S,
+                      maxRetries: Option[Int],
+                      backoff: Backoff = SimpleExponentialBackoff(5.seconds, 10.seconds, 1.1D),
+                      isTransient: Throwable => Boolean = throwableToFalse,
+                      isFatal: Throwable => Boolean = throwableToFalse,
+                      onRetry: (Throwable, S) => S = noOpOnRetry)
+                     (implicit timer: Timer[IO], statefulIoException: StatefulIoError[S]): IO[A] = {
+    lazy val delay = backoff.backoffMillis.millis
+
+    def fail(throwable: Throwable) = IO.raiseError(statefulIoException.toThrowable(state, throwable))
+
+    io handleErrorWith {
+      case throwable if isFatal(throwable) => fail(throwable)
+      case throwable =>
```

Comment:
Stop catching `java.lang.Error`.

```scala
io handleErrorWith {
      case exception: Exception if isFatal(exception) => fail(exception)
      case exception: Exception => ...
      case throwable => throw throwable
}
```

We should never be trying to do extra biz logic on fatal JVM errors such as `java.lang.OutOfMemoryError` etc. Also change the contracts/signatures to use `java.lang.Exception`, for example `isFatal: Exception => Boolean`.

Reply:
For posterity:

There's a bit of debate on the internet. Different camps agree that one shouldn't catch all `java.lang.Throwable`. However a scala-based camp prefers the use of `NonFatal` and a java-based camp is all about checked-vs-unchecked `Exception` handling.

I don't know what's "right" at the moment, but will accept any patches that don't catch OOME among others.